src/tnc.rs

```diff
@@ -163,28 +165,86 @@ impl Tnc {
             ConnectConfig::TcpKiss(config) => Box::new(TcpKissTnc::open(&config)?),
             ConnectConfig::LinuxIf(config) => Box::new(LinuxIfTnc::open(&config)?),
         };
-        Ok(Self { imp })
+        Ok(Tnc(Arc::new(Mutex::new(TncInner::new(imp)))))
     }
 
     /// Transmit a frame on the radio. Transmission is not guaranteed even if a
     /// `Ok` result is returned.
     pub fn send_frame(&self, frame: &Ax25Frame) -> Result<(), TncError> {
-        self.imp.send_frame(frame)
+        self.0.lock().unwrap().send_frame(frame)
     }
 
-    /// Block to receive a frame from the radio. If you want to do this on
-    /// a separate thread from sending, clone the `Tnc`.
-    pub fn receive_frame(&self) -> Result<Ax25Frame, TncError> {
-        self.imp.receive_frame()
+    /// Create a new `Receiver<Result<Ax25Frame, TncError>>`
+    /// This will receive a copy of all incoming frames.
+    pub fn incoming(&mut self) -> Receiver<Ax25FrameResult> {
+        self.0.lock().unwrap().incoming()
     }
 }
 
-impl Clone for Tnc {
-    fn clone(&self) -> Self {
-        Tnc {
-            imp: self.imp.clone(),
+pub type Ax25FrameResult = Result<Ax25Frame, Arc<TncError>>;
+
+struct TncInner {
+    imp: Box<dyn TncImpl>,
+    senders: Arc<Mutex<Vec<Sender<Ax25FrameResult>>>>,
+    disconnected: Arc<AtomicBool>,
+}
+
+impl TncInner {
+    fn new(imp: Box<dyn TncImpl>) -> Self {
+        let senders: Arc<Mutex<Vec<Sender<Ax25FrameResult>>>> = Arc::new(Mutex::new(Vec::new()));
+        let disconnected = Arc::new(AtomicBool::new(false));
+
+        {
+            let imp = imp.clone();
+            let senders = senders.clone();
+            let disconnected = disconnected.clone();
+
+            thread::spawn(move || {
+                while !disconnected.load(Ordering::Relaxed) {
+                    let x = match imp.receive_frame() {
+                        Ok(a) => Ok(a),
+                        Err(e) => Err(Arc::new(e)),
```

Comment:
Ah I didn't foresee that issue. A bit annoying but this is okay until I dream up something else.